integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+     * {@code null} {@link Xid}
+     *
+     * @param exceptionConverter a {@link ExceptionConverter} that accepts a {@link XARoutine} and a {@link
+     * SQLException} and converts the {@link SQLException} to an <em>appropriate</em> {@link XAException} following the
+     * rules defined by the <a href="https://pubs.opengroup.org/onlinepubs/009680699/toc.pdf">XA Specification</a> as
+     * interpreted by the specification of the {@code javax.transaction.xa} package and its classes; may be {@code null}
+     * in which case a default implementation will be used instead
+     *
+     * @see #start(Xid, int)
+     */
+    LocalXAResource(Function<? super Xid, ? extends Connection> connectionFunction, ExceptionConverter exceptionConverter) {
+        super();
+        this.connectionFunction = Objects.requireNonNull(connectionFunction, "connectionFunction");
+        this.exceptionConverter = exceptionConverter == null ? LocalXAResource::convert0 : exceptionConverter;
+    }
+
+
+    /*
+     * Instance methods.
+     */
+
+
+    @Override // XAResource
+    public void start(Xid xid, int flags) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "start", new Object[] {xid, flagsToString(flags)});
+        }
+        requireNonNullXid(xid);
+
+        BiFunction<Xid, Association, Association> remappingFunction;
+        switch (flags) {
+        case TMJOIN:
+            remappingFunction = LocalXAResource::join;
+            break;
+        case TMNOFLAGS:
+            remappingFunction = this::start;
+            break;
+        case TMRESUME:
+            remappingFunction = LocalXAResource::resume;
+            break;
+        default:
+            // Bad flags.
+            throw (XAException) new XAException(XAER_INVAL)
+                .initCause(new IllegalArgumentException("xid: " + xid + "; flags: " + flagsToString(flags)));
+        }
+
+        this.computeAssociation(XARoutine.START, xid, remappingFunction);
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "start");
+        }
+    }
+
+    // (Remapping BiFunction, used in start() above and supplied to computeAssociation() below.)
+    private Association start(Xid x, Association a) {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "start", new Object[] {x, a});
+        }
+        assert x != null; // x has already been vetted and is known to be non-null
+        if (a != null) {
+            throw new UncheckedXAException((XAException) new XAException(XAER_DUPID)
+                                           .initCause(new IllegalArgumentException("xid: " + x + "; association: " + a)));
+        }
+        Connection c;
+        try {
+            c = this.connectionFunction.apply(x);
+        } catch (RuntimeException e) {
```

Comment:
Figuring out whether to throw an `XAException` with `XAER_RMERR` or `XAER_RMFAIL` is a problem that has resulted in dozens of bugs by all major `XAResource` implementors over a couple of decades. It is far from straightforward and ends up being a _de facto_ decision: you have to do what Narayana expects, regardless of whether it implements the specification properly or not, because it is the only real robust JTA implementation out there and has encountered just about everything in its 30+ years of existence.